src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -1000,603 +946,315 @@ private ImmutableArray<Symbol> CreateMembers()
+            private ImmutableArray<FieldSymbol> _lazyDefaultElementFields;
+            private SmallDictionary<Symbol, Symbol>? _lazyUnderlyingDefinitionToMemberMap;
 
-        public override TypeKind TypeKind
-        {
-            get
-            {
-                // From the language perspective tuple is a value type 
-                // composed of its underlying elements
-                return TypeKind.Struct;
-            }
-        }
+            /// <summary>
+            /// The same named type, but without element names.
+            /// </summary>
+            internal NamedTypeSymbol TupleUnderlyingType { get; }
 
-        public override Symbol ContainingSymbol
-        {
-            get
+            internal TupleUncommonData(NamedTypeSymbol underlyingType)
             {
-                return _underlyingType.ContainingSymbol;
-            }
-        }
+                Debug.Assert(underlyingType is object);
+                Debug.Assert(underlyingType.IsTupleType);
+                Debug.Assert(underlyingType.TupleElementNames.IsDefault);
 
-        public override ImmutableArray<Location> Locations
-        {
-            get
-            {
-                return _locations;
+                TupleUnderlyingType = underlyingType;
+                Locations = ImmutableArray<Location>.Empty;
             }
-        }
 
-        public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences
-        {
-            get
+            internal TupleUncommonData(NamedTypeSymbol underlyingType, ImmutableArray<string> elementNames,
+                ImmutableArray<Location> elementLocations, ImmutableArray<bool> errorPositions, ImmutableArray<Location> locations)
+                : this(underlyingType)
             {
-                return GetDeclaringSyntaxReferenceHelper<CSharpSyntaxNode>(_locations);
+                ElementNames = elementNames;
+                ElementLocations = elementLocations;
+                ErrorPositions = errorPositions;
+                Locations = locations.IsDefault ? ImmutableArray<Location>.Empty : locations;
             }
-        }
 
-        internal override bool Equals(TypeSymbol t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt = null)
-        {
-            if ((comparison & TypeCompareKind.IgnoreTupleNames) != 0)
+            internal bool EqualsIgnoringTupleUnderlyingType(TupleUncommonData? other)
             {
-                if (t2?.IsTupleType == true)
+                if (other is null && this.ElementNames.IsDefault && this.ElementLocations.IsDefault && this.ErrorPositions.IsDefault)
                 {
-                    t2 = t2.TupleUnderlyingType;
+                    return true;
                 }
 
-                return _underlyingType.Equals(t2, comparison, isValueTypeOverrideOpt);
-            }
+                return other is object &&
+                    this.ElementNames == other.ElementNames
```

Comment:
>this.ElementNames == other.ElementNames [](start = 20, length = 39)

I don't think this compares element-wise. Is that the intent?

Reply:
It does compare element-wise:
```
        // Summary:
        //     Returns a value that indicates if two arrays are equal.
        // Returns:
        //     true if the arrays are equal; otherwise, false.
        [NonVersionableAttribute]
        public static bool operator ==(ImmutableArray<T> left, ImmutableArray<T> right);
```

---
In reply to: [355728188](https://github.com/dotnet/roslyn/pull/39370#discussion_r355728188) [](ancestors = 355728188)

Reply:
`ImmutableArray.Create(1) == ImmutableArray.Create(1)` is `false`


---
In reply to: [355757723](https://github.com/dotnet/roslyn/pull/39370#discussion_r355757723) [](ancestors = 355757723,355728188)

Reply:
Hum... Not sure what is going on. I'll address just to be sure.

---
In reply to: [355836372](https://github.com/dotnet/roslyn/pull/39370#discussion_r355836372) [](ancestors = 355836372,355757723,355728188)